File: Searching-Array/prg.cpp
#include <iostream>
using namespace std;

void linearSearch(int arr[], int size, int key) {
    for (int i = 0; i < size; i++)
        if (arr[i] == key) {
            cout << "Found at index " << i << endl;
            return;
        }
    cout << "Not found" << endl;
}

void bubbleSort(int arr[], int size) {
    for (int i = 0; i < size - 1; i++)
        for (int j = 0; j < size - i - 1; j++)
            if (arr[j] > arr[j + 1])
                swap(arr[j], arr[j + 1]);
    cout << "Sorted Array: ";
    for (int i = 0; i < size; i++) cout << arr[i] << " ";
    cout << endl;
}

int interpolationSearch(int arr[], int size, int key) {
    int low = 0, high = size - 1;
    while (low <= high && key >= arr[low] && key <= arr[high]) {
        int pos = low + ((key - arr[low]) * (high - low) / (arr[high] - arr[low]));
        if (arr[pos] == key) return pos;
        if (arr[pos] < key) low = pos + 1;
        else high = pos - 1;
    }
    return -1;
}

int main() {
    int arr[] = {10, 20, 30, 40, 50};
    int size = sizeof(arr) / sizeof(arr[0]);
    int choice, key;

    while (true) {
        cout << "1. Linear Search\n2. Bubble Sort\n3. Interpolation Search\n4. Exit\n";
        cin >> choice;
        if (choice == 4) break;

        if (choice == 1 || choice == 3) {
            cout << "Enter key to search: ";
            cin >> key;
            if (choice == 1) linearSearch(arr, size, key);
            else cout << "Found at index: " << interpolationSearch(arr, size, key) << endl;
        } else if (choice == 2) {
            bubbleSort(arr, size);
        }
    }
    return 0;
}

File: Searching-Array/prg.rs
use std::io;

fn linear_search(arr: &[i32], key: i32) {
    for (i, &val) in arr.iter().enumerate() {
        if val == key {
            println!("Found at index {}", i);
            return;
        }
    }
    println!("Not found");
}

fn bubble_sort(arr: &mut Vec<i32>) {
    let n = arr.len();
    for i in 0..n {
        for j in 0..n - i - 1 {
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
            }
        }
    }
    println!("Sorted Array: {:?}", arr);
}

fn interpolation_search(arr: &[i32], key: i32) -> i32 {
    let mut low = 0;
    let mut high = arr.len() - 1;
    while low <= high && key >= arr[low] && key <= arr[high] {
        let pos = low + (((key - arr[low]) * (high - low) as i32) / (arr[high] - arr[low])) as usize;
        if arr[pos] == key {
            return pos as i32;
        }
        if arr[pos] < key {
            low = pos + 1;
        } else {
            high = pos - 1;
        }
    }
    -1
}

fn main() {
    let mut arr = vec![10, 20, 30, 40, 50];
    let mut input = String::new();

    loop {
        println!("1. Linear Search\n2. Bubble Sort\n3. Interpolation Search\n4. Exit");
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let choice: i32 = input.trim().parse().unwrap();

        if choice == 4 {
            break;
        }

        if choice == 1 || choice == 3 {
            println!("Enter key to search: ");
            input.clear();
            io::stdin().read_line(&mut input).unwrap();
            let key: i32 = input.trim().parse().unwrap();

            if choice == 1 {
                linear_search(&arr, key);
            } else {
                let result = interpolation_search(&arr, key);
                println!("Found at index: {}", result);
            }
        } else if choice == 2 {
            bubble_sort(&mut arr);
        }
    }
}
